python/distributed-ucxx/distributed_ucxx/ucxx.py

```diff
@@ -90,13 +90,84 @@ def synchronize_stream(stream=0):
     stream.synchronize()
 
 
+def _stop_notifier_thread_and_progress_tasks():
+    """Stop the notifier thread and progress tasks.
+
+    If no Dask resources that make use of UCXX communicator are running anymore,
+    stop the notifier thread and progress tasks to allow for clean shutdown.
+    """
+    ctx = ucxx.core._get_ctx()
+    if len(ctx._dask_resources) == 0:
+        ctx.stop_notifier_thread()
+        ctx.progress_tasks.clear()
+
+
+def _register_dask_resource(resource):
+    """Register a Dask resource with the UCXX context.
+
+    Register a Dask resource with the UCXX context to keep track of it, so that
+    the notifier thread and progress tasks may be stopped when no more resources
+    """
+    ctx = ucxx.core._get_ctx()
+
+    with ctx._dask_resources_lock:
+        ctx._dask_resources.add(resource)
+        ctx.start_notifier_thread()
+        ctx.continuous_ucx_progress()
```

Comment:
A little confused here. AIUI, `ucxx.core._get_ctx()` returns a global `ApplicationContext` object. And `ctx.start_notifier_thread` creates a new thread in the application context, overwriting any existing running thread.

So does this mean it's not really "safe" to create more than one "dask resource" per process (and hence per application context). Or does it not matter that we throw away the `ctx.notifier_thread` and replace it with a new one when we call `start_notifier_thread`?